flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java

```diff
@@ -0,0 +1,340 @@
+	private static final String TASK_MANAGER_POD_FORMAT = "%s-taskmanager-%d-%d";
+
+	private final Map<ResourceID, KubernetesWorkerNode> workerNodes = new HashMap<>();
+
+	private final double defaultCpus;
+
+	/** When ResourceManager failover, the max attempt should recover. */
+	private final AtomicLong currentMaxAttemptId = new AtomicLong(0);
+
+	/** Current max pod index. When creating a new pod, it should increase one. */
+	private final AtomicLong currentMaxPodId = new AtomicLong(0);
+
+	private final String clusterId;
+
+	private final FlinkKubeClient kubeClient;
+
+	private final ContaineredTaskManagerParameters taskManagerParameters;
+
+	private final List<String> taskManagerStartCommand;
+
+	/** The number of pods requested, but not yet granted. */
+	private int numPendingPodRequests = 0;
+
+	public KubernetesResourceManager(
+			RpcService rpcService,
+			String resourceManagerEndpointId,
+			ResourceID resourceId,
+			Configuration flinkConfig,
+			HighAvailabilityServices highAvailabilityServices,
+			HeartbeatServices heartbeatServices,
+			SlotManager slotManager,
+			JobLeaderIdService jobLeaderIdService,
+			ClusterInformation clusterInformation,
+			FatalErrorHandler fatalErrorHandler,
+			ResourceManagerMetricGroup resourceManagerMetricGroup) {
+		super(
+			flinkConfig,
+			System.getenv(),
+			rpcService,
+			resourceManagerEndpointId,
+			resourceId,
+			highAvailabilityServices,
+			heartbeatServices,
+			slotManager,
+			jobLeaderIdService,
+			clusterInformation,
+			fatalErrorHandler,
+			resourceManagerMetricGroup);
+		this.clusterId = flinkConfig.getString(KubernetesConfigOptions.CLUSTER_ID);
+		this.defaultCpus = flinkConfig.getDouble(KubernetesConfigOptions.TASK_MANAGER_CPU, numSlotsPerTaskManager);
+
+		this.kubeClient = createFlinkKubeClient();
+
+		this.taskManagerParameters =
+			ContaineredTaskManagerParameters.create(flinkConfig, taskExecutorResourceSpec, numSlotsPerTaskManager);
+
+		this.taskManagerStartCommand = getTaskManagerStartCommand();
+	}
+
+	@Override
+	protected Configuration loadClientConfiguration() {
+		return GlobalConfiguration.loadConfiguration();
+	}
+
+		@Override
+	protected void initialize() throws ResourceManagerException {
+		recoverWorkerNodesFromPreviousAttempts();
+
+		kubeClient.watchPodsAndDoCallback(getTaskManagerLabels(), this);
+	}
+
+	@Override
+	public CompletableFuture<Void> onStop() {
+		// shut down all components
+		Throwable exception = null;
+
+		try {
+			kubeClient.close();
+		} catch (Throwable t) {
+			exception = t;
+		}
+
+		return getStopTerminationFutureOrCompletedExceptionally(exception);
+	}
+
+	@Override
+	protected void internalDeregisterApplication(ApplicationStatus finalStatus, @Nullable String diagnostics) {
+		LOG.info(
+			"Stopping kubernetes cluster, clusterId: {}, diagnostics: {}",
+			clusterId,
+			diagnostics == null ? "" : diagnostics);
+		kubeClient.stopAndCleanupCluster(clusterId);
```

Comment:
Do we also need call this `Utils.deleteApplicationFiles(taskManagerParameters.taskManagerEnv())` as done in `YarnResourceManager`?

Reply:
We do not upload any files to dfs staging directory. So `Utils.deleteApplicationFiles` is unnecessary.